File: src/go/radio/api_scene.go
package radio

import "time"

type SceneApi struct {
	*ApiModule
}

func NewSceneApi(rm *RadioManager) *SceneApi {
	return &SceneApi{NewApiModule(rm)}
}

func (api *SceneApi) ListAllMeshObjectInScene() interface{} {
	return api.Send("scene.listAllMeshObjectInScene", nil, "", 3*time.Second, nil)
}

func (api *SceneApi) GetObjectInfoByID(objectId string) interface{} {
	return api.Send("scene.getObjectInfoById", map[string]interface{}{"objectId": objectId}, "", 3*time.Second, nil)
}

func (api *SceneApi) RemoveObjectByID(objectId string) interface{} {
	return api.Send("scene.removeObjectById", map[string]interface{}{"objectId": objectId}, "", 3*time.Second, nil)
}

func (api *SceneApi) MoveObjectByID(objectId string, position XYZ) interface{} {
	return api.Send("scene.moveObjectById", map[string]interface{}{
		"objectId": objectId,
		"position": []float64{position[0], position[1], position[2]},
	}, "", 3*time.Second, nil)
}

func (api *SceneApi) SetObjectRadioMaterial(objectId string, materialId *string, thicknessM *float64) interface{} {
	return api.Send("scene.setObjectRadioMaterial", map[string]interface{}{
		"objectId":    objectId,
		"materialId":  materialId,
		"thickness_m": thicknessM,
	}, "", 3*time.Second, nil)
}

func (api *SceneApi) UpdateMeshViewMaterial(meshId string, cmd ViewMaterialChangeCommand) interface{} {
	return api.Send("scene.updateMeshViewMaterial", map[string]interface{}{
		"meshId":                    meshId,
		"viewMaterialChangeCommand": cmd,
	}, "", 3*time.Second, nil)
}

func (api *SceneApi) UpdateMeshViewMaterialSimple(meshId string, cmd ViewMaterialChangeCommandSimple) interface{} {
	return api.Send("scene.updateMeshViewMaterialSimple", map[string]interface{}{
		"meshId":                          meshId,
		"viewMaterialChangeCommandSimple": cmd,
	}, "", 3*time.Second, nil)
}

File: src/lua/radio/api_scene.lua
-- SceneApi implementation in Lua

local api_module = require("radio/api_module")

local SceneApi = {}
SceneApi.__index = SceneApi
setmetatable(SceneApi, { __index = api_module.ApiModule })

function SceneApi.new(rm)
    local self = api_module.ApiModule.new(rm)
    setmetatable(self, SceneApi)
    return self
end

function SceneApi:listAllMeshObjectInScene()
    return self:send("scene.listAllMeshObjectInScene")
end

function SceneApi:getObjectInfoById(objectId)
    return self:send("scene.getObjectInfoById", { objectId = objectId })
end

function SceneApi:removeObjectById(objectId)
    return self:send("scene.removeObjectById", { objectId = objectId })
end

function SceneApi:moveObjectById(objectId, position)
    return self:send("scene.moveObjectById", {
        objectId = objectId,
        position = { position[1], position[2], position[3] }
    })
end

function SceneApi:setObjectRadioMaterial(objectId, materialId, thickness_m)
    return self:send("scene.setObjectRadioMaterial", {
        objectId = objectId,
        materialId = materialId,
        thickness_m = thickness_m
    })
end

function SceneApi:updateMeshViewMaterial(meshId, viewMaterialChangeCommand)
    return self:send("scene.updateMeshViewMaterial", {
        meshId = meshId,
        viewMaterialChangeCommand = viewMaterialChangeCommand
    })
end

function SceneApi:updateMeshViewMaterialSimple(meshId, viewMaterialChangeCommandSimple)
    return self:send("scene.updateMeshViewMaterialSimple", {
        meshId = meshId,
        viewMaterialChangeCommandSimple = viewMaterialChangeCommandSimple
    })
end

return {
    SceneApi = SceneApi
}
